File: packages/twerk-slots/src/twerk_slots/cli/slot/gc.py
"""`slot gc` — free slots whose branch has a merged or closed PR."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated

import click

from twerk_core import get_console
from twerk_core.clinkr.dataclass_json import JsonSerializable
from twerk_core.clinkr.ensure import Ensure
from twerk_core.clinkr.exit import ClinkrExit
from twerk_core.clinkr.operation import clinkr_operation
from twerk_core.gh.types import PRState
from twerk_slots.cli.slot.context import load_slots_context
from twerk_slots.gc import (
    SlotGcAction,
    SlotGcOutcome,
    execute_gc_plan,
    outcome_from_plan,
    plan_gc,
)
from twerk_slots.repo_context import NoGitRepo


@dataclass(frozen=True)
class SlotGcRequest:
    dry_run: Annotated[bool, click.Option(["--dry-run"], is_flag=True, default=False)] = False
    force: Annotated[bool, click.Option(["-f", "--force"], is_flag=True, default=False)] = False


@dataclass(frozen=True)
class SlotGcResultEntry:
    slot_name: str
    branch_name: str
    worktree_path: str
    action: SlotGcAction
    pr_number: int | None
    pr_state: PRState | None
    pr_url: str | None
    message: str | None


@dataclass(frozen=True)
class SlotGcResult(JsonSerializable):
    entries: tuple[SlotGcResultEntry, ...]
    freed_count: int
    kept_count: int
    skipped_count: int
    error_count: int
    dry_run: bool
    cancelled: bool = False


_ACTION_LABELS: dict[SlotGcAction, tuple[str, str]] = {
    "freed": ("[green]✓ freed[/green]", "green"),
    "would_free": ("[yellow]→ would free[/yellow]", "yellow"),
    "kept_open_pr": ("[blue]• kept (open PR)[/blue]", "blue"),
    "kept_no_pr": ("[dim]• kept (no PR)[/dim]", "dim"),
    "skipped_dirty": ("[yellow]! skipped (dirty)[/yellow]", "yellow"),
    "error": ("[red]✗ error[/red]", "red"),
}


def render_slot_gc(result: SlotGcResult) -> None:
    console = get_console()
    if result.cancelled:
        console.print("[yellow]Cancelled — no slots freed.[/yellow]")
        return
    if not result.entries:
        console.print("[dim]No assignments to sweep.[/dim]")
        return

    for entry in result.entries:
        label, _colour = _ACTION_LABELS[entry.action]
        pr_suffix = ""
        if entry.pr_number is not None:
            pr_suffix = f" [dim]PR #{entry.pr_number} {entry.pr_state}[/dim]"
        console.print(
            f"{label} [bold cyan]{entry.slot_name}[/bold cyan] "
            f"([yellow]{entry.branch_name}[/yellow]){pr_suffix}"
        )
        if entry.message:
            console.print(f"    [dim]{entry.message}[/dim]")

    verb = "Would free" if result.dry_run else "Freed"
    console.print(
        f"\n[bold]{verb} {result.freed_count}[/bold]; "
        f"kept {result.kept_count}; "
        f"skipped {result.skipped_count}; "
        f"errors {result.error_count}"
    )


def _result_from_outcome(outcome: SlotGcOutcome, *, cancelled: bool = False) -> SlotGcResult:
    return SlotGcResult(
        entries=tuple(
            SlotGcResultEntry(
                slot_name=e.slot_name,
                branch_name=e.branch_name,
                worktree_path=str(e.worktree_path),
                action=e.action,
                pr_number=e.pr_number,
                pr_state=e.pr_state,
                pr_url=e.pr_url,
                message=e.message,
            )
            for e in outcome.entries
        ),
        freed_count=outcome.freed_count,
        kept_count=outcome.kept_count,
        skipped_count=outcome.skipped_count,
        error_count=outcome.error_count,
        dry_run=outcome.dry_run,
        cancelled=cancelled,
    )


@clinkr_operation(
    name="gc",
    help="Free slots whose branch has a merged or closed PR.",
    human_renderer=render_slot_gc,
)
def run_slot_gc(ctx: click.Context, request: SlotGcRequest) -> ClinkrExit[SlotGcResult]:
    slots_ctx = load_slots_context(ctx)
    if isinstance(slots_ctx, NoGitRepo):
        Ensure.fail(error_type="not_in_repo", message=slots_ctx.message)

    Ensure.true(
        slots_ctx.pool_state.exists(),
        error_type="pool_empty",
        message="No pool configured. Run `slot checkout` first.",
    )

    Ensure.true(
        not (request.dry_run and request.force),
        error_type="conflicting_flags",
        message="--dry-run and --force are mutually exclusive.",
    )

    plan = plan_gc(slots_ctx)

    if request.dry_run:
        return ClinkrExit.ok(_result_from_outcome(outcome_from_plan(plan, dry_run=True)))

    if plan.would_free_count == 0:
        return ClinkrExit.ok(_result_from_outcome(outcome_from_plan(plan, dry_run=False)))

    if request.force:
        return ClinkrExit.ok(_result_from_outcome(execute_gc_plan(slots_ctx, plan)))

    preview = _result_from_outcome(outcome_from_plan(plan, dry_run=True))
    render_slot_gc(preview)
    proceed = click.confirm(
        f"Free {plan.would_free_count} slot(s)?",
        default=False,
    )
    if proceed:
        return ClinkrExit.ok(_result_from_outcome(execute_gc_plan(slots_ctx, plan)))
    return ClinkrExit.ok(
        _result_from_outcome(outcome_from_plan(plan, dry_run=False), cancelled=True)
    )

File: ts/packages/capabilities/slot/src/operations/gc.ts
import {
	failure,
	negative,
	ok,
	requireInteractiveOrUsageError,
	resolveRenderCapabilities,
	type Caps,
	type RenderCapabilities,
} from "@sdl/clinkr";
import { cell, dim, paint, renderTable, stripAnsiWhenDisabled } from "@sdl/core/cli-theme";
import { optionalEntry } from "@sdl/core/primitives";
import { z } from "zod";

import type { RepoSlotContext, SlotCliContext } from "../context.ts";
import {
	outcomeFromGcPlan,
	planGc,
	planGcCleanup,
	executeGcPlan,
	type SlotGcOutcome,
} from "../lifecycle/gc.ts";
import type { SlotFreeCleanupAction } from "../lifecycle/release-cleanup.ts";
import { renderCleanupLines } from "./cleanup-rendering.ts";
import {
	buildSlotDestructiveResultBlock,
	renderSlotDestructiveResultBlock,
} from "./destructive-presentation.ts";
import { cleanupSchema } from "./result-schemas.ts";
const gcEntrySchema = z.object({
	slotName: z.string(),
	branchName: z.string(),
	worktreePath: z.string(),
	action: z.union([
		z.literal("freed"),
		z.literal("would-free"),
		z.literal("kept-open-pr"),
		z.literal("kept-no-pr"),
		z.literal("skipped-dirty"),
		z.literal("skipped-operation"),
		z.literal("error"),
	]),
	prNumber: z.number().int().nullable(),
	prState: z.union([z.literal("OPEN"), z.literal("CLOSED"), z.literal("MERGED")]).nullable(),
	prUrl: z.string().nullable(),
	message: z.string().nullable(),
	cleanup: z.array(cleanupSchema),
});

export const gcRequestSchema = z.object({
	dryRun: z.boolean().default(false).describe("Preview without mutating."),
	force: z.boolean().default(false).describe("Free candidates without prompting."),
	deleteBranches: z.boolean().default(false).describe("Delete local branches for freed slots."),
});

export const gcResultSchema = z.object({
	entries: z.array(gcEntrySchema),
	freedCount: z.number().int(),
	keptCount: z.number().int(),
	skippedCount: z.number().int(),
	errorCount: z.number().int(),
	dryRun: z.boolean(),
	cleanupErrorCount: z.number().int(),
	cancelled: z.boolean().optional(),
});

export type GcRequest = z.infer<typeof gcRequestSchema>;
export type GcResult = z.infer<typeof gcResultSchema>;

export async function runGc(ctx: SlotCliContext, request: GcRequest) {
	if (ctx.repo.type !== "repo") return failure(ctx.repo.errorType, ctx.repo.message);
	if (request.dryRun && request.force)
		return failure("conflicting-flags", "--dry-run and --force cannot be combined.");
	const repoCtx: RepoSlotContext = { ...ctx, repo: ctx.repo };
	const cleanupActions: readonly SlotFreeCleanupAction[] = request.deleteBranches
		? ["local-branch"]
		: [];
	const plan = await planGc(repoCtx);
	if (plan.type === "failure") return failure(plan.failure.errorType, plan.failure.message);
	if (request.dryRun) {
		const cleanup = await planGcCleanup(repoCtx, plan.outcome, cleanupActions);
		return ok(toGcResult(outcomeFromGcPlan(plan.outcome, { isDryRun: true, cleanup })));
	}
	if (plan.outcome.wouldFreeCount === 0)
		return ok(toGcResult(outcomeFromGcPlan(plan.outcome, { isDryRun: false })));
	if (!request.force) {
		const gate = requireInteractiveOrUsageError(ctx.interaction, {
			message: "Destructive gc requires --force when non-interactive (or run --dry-run first).",
			missingFlag: "--force",
			howToSupply: "Pass --force (or -f) to free slots without prompting, or run --dry-run first.",
		});
		if (gate) return gate;
		const cleanup = await planGcCleanup(repoCtx, plan.outcome, cleanupActions);
		const renderCapabilities = repoCtx.renderCapabilities;
		const preview = renderGc(
			toGcResult(outcomeFromGcPlan(plan.outcome, { isDryRun: true, cleanup })),
			renderCapabilities,
		);
		repoCtx.stderr(`${stripAnsiWhenDisabled(preview, renderCapabilities)}\n`);
		const accepted = await repoCtx.interaction.confirm({
			message: confirmationMessage(plan.outcome.wouldFreeCount, {
				shouldDeleteBranches: request.deleteBranches,
			}),
			defaultAnswer: "yes",
		});
		if (accepted.type === "aborted") return failure("aborted", "Aborted!");
		if (accepted.type === "declined")
			return ok(
				toGcResult(outcomeFromGcPlan(plan.outcome, { isDryRun: false, cleanup }), {
					isCancelled: true,
				}),
			);
	}
	const outcome = await executeGcPlan(repoCtx, plan.outcome, { cleanupActions });
	const result = toGcResult(outcome);
	if (outcome.cleanupErrorCount > 0)
		return negative("Slot gc completed with cleanup errors.", {
			data: result,
			human: renderGc(result),
		});
	return ok(result);
}

export function renderGc(
	result: GcResult,
	caps: RenderCapabilities = { canEmitAnsi: false },
): string {
	const resolvedCaps = resolveRenderCapabilities(caps);
	const body = renderGcDetails(result, resolvedCaps);
	return renderSlotDestructiveResultBlock(
		caps,
		buildSlotDestructiveResultBlock({
			kind: gcResultKind(result),
			headline: gcHeadline(result),
			...optionalEntry("body", body),
		}),
	);
}

function gcResultKind(result: GcResult): "success" | "failure" | "refusal" {
	if (result.cancelled === true) return "refusal";
	if (result.cleanupErrorCount > 0 || result.errorCount > 0) return "failure";
	return "success";
}

function gcHeadline(result: GcResult): string {
	if (result.cancelled === true) return "Cancelled slot gc.";
	if (result.cleanupErrorCount > 0) return "Slot gc completed with cleanup errors.";
	if (result.errorCount > 0) return "Slot gc completed with errors.";
	if (result.dryRun) {
		if (result.freedCount === 0) return "No slots would be freed.";
		return `Would free ${result.freedCount} slot(s).`;
	}
	if (result.freedCount === 0) return "No slots freed.";
	return `Freed ${result.freedCount} slot(s).`;
}

function renderGcDetails(result: GcResult, caps: Caps): string | undefined {
	if (result.entries.length === 0) return undefined;
	const tableRows = result.entries.map((entry) => [
		gcActionCell(caps, entry.action),
		cell(paint(caps, "accent", entry.slotName), entry.slotName),
		cell(entry.branchName),
		gcPrCell(caps, entry),
	]);
	const tableLines = renderTable({
		caps,
		columns: [
			{ header: "ACTION", width: "auto" },
			{ header: "SLOT", width: "auto" },
			{ header: "BRANCH", width: "auto" },
			{ header: "PR", width: "auto" },
		],
		rows: tableRows,
	});
	const lines: string[] = [tableLines[0] ?? ""];
	const rowLines = tableLines.slice(1);
	result.entries.forEach((entry, index) => {
		const rowLine = rowLines[index];
		if (rowLine !== undefined) lines.push(rowLine);
		if (entry.message !== null) lines.push(`  ${paint(caps, "muted", "note:")} ${entry.message}`);
		lines.push(...renderGcCleanupDetails(caps, entry, { isDryRun: result.dryRun }));
	});
	lines.push(gcSummaryLine(result));
	return lines.join("\n");
}

function gcActionCell(caps: Caps, action: GcResult["entries"][number]["action"]) {
	const text = gcActionText(action);
	return cell(paint(caps, gcActionIntent(action), text), text);
}

function gcPrCell(caps: Caps, entry: GcResult["entries"][number]) {
	const text = prText(entry);
	if (entry.prState === null) return cell(paint(caps, "muted", text), text);
	return cell(paint(caps, prStateIntent(entry.prState), text), text);
}

function renderGcCleanupDetails(
	caps: Caps,
	entry: GcResult["entries"][number],
	options: { isDryRun: boolean },
): readonly string[] {
	return renderCleanupLines(entry.cleanup, { isDryRun: options.isDryRun, caps }).map(
		(line) => `  ${paint(caps, "muted", "cleanup:")} ${line}`,
	);
}

function gcSummaryLine(result: GcResult): string {
	const freedLabel = result.dryRun ? "would free" : "freed";
	let summary = `${freedLabel} ${result.freedCount}; kept ${result.keptCount}; skipped ${result.skippedCount}; errors ${result.errorCount}`;
	if (result.cleanupErrorCount > 0)
		summary = `${summary}; cleanup errors ${result.cleanupErrorCount}`;
	return dim(`Summary: ${summary}`);
}

function gcActionText(action: GcResult["entries"][number]["action"]): string {
	switch (action) {
		case "freed":
			return "Freed";
		case "would-free":
			return "Would free";
		case "kept-open-pr":
			return "Kept: open PR";
		case "kept-no-pr":
			return "Kept: no PR";
		case "skipped-dirty":
			return "Skipped: dirty";
		case "skipped-operation":
			return "Skipped: operation";
		case "error":
			return "Error";
	}
}

function gcActionIntent(
	action: GcResult["entries"][number]["action"],
): "success" | "warn" | "error" | "accent" | "muted" {
	switch (action) {
		case "freed":
			return "success";
		case "would-free":
			return "accent";
		case "kept-open-pr":
			return "warn";
		case "kept-no-pr":
		case "skipped-dirty":
		case "skipped-operation":
			return "muted";
		case "error":
			return "error";
	}
}

function prStateIntent(
	state: NonNullable<GcResult["entries"][number]["prState"]>,
): "success" | "warn" | "muted" {
	if (state === "OPEN") return "warn";
	if (state === "MERGED") return "success";
	return "muted";
}

function prText(entry: GcResult["entries"][number]): string {
	if (entry.prNumber === null) return "—";
	return `#${entry.prNumber} ${entry.prState}`;
}

function confirmationMessage(count: number, options: { shouldDeleteBranches: boolean }): string {
	if (options.shouldDeleteBranches)
		return `Free ${count} merged/closed slot(s) and force-delete their local branches?`;
	return `Free ${count} merged/closed slot(s)?`;
}

function toGcResult(outcome: SlotGcOutcome, options: { isCancelled?: boolean } = {}): GcResult {
	return {
		entries: outcome.entries.map((entry) => ({ ...entry, cleanup: [...entry.cleanup] })),
		freedCount: options.isCancelled === true ? 0 : outcome.freedCount,
		keptCount: outcome.keptCount,
		skippedCount: outcome.skippedCount,
		errorCount: outcome.errorCount,
		dryRun: outcome.dryRun,
		cleanupErrorCount: outcome.cleanupErrorCount,
		...optionalEntry("cancelled", options.isCancelled),
	};
}
